src/MudletInstanceCoordinator.cpp

```diff
@@ -0,0 +1,136 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
+ ***************************************************************************/
+
+#include "MudletInstanceCoordinator.h"
+#include "Host.h"
+#include "mudlet.h"
+#include <QLocalSocket>
+
+const int WAIT_FOR_RESPONSE_MS = 500;
+
+MudletInstanceCoordinator::MudletInstanceCoordinator(const QString& serverName, QObject* parent) : QLocalServer(parent), mServerName(serverName) {}
+
+void MudletInstanceCoordinator::queuePackage(const QString& packageName)
+{
+    mQueuedPackagePaths << packageName;
+}
+
+// Install the package queue on another instance of Mudlet.
+// Returns true on success
+bool MudletInstanceCoordinator::installPackagesRemotely()
+{
+    // Pass the absolute path of the package to the active Mudlet Server
+    // The Mudlet Server may be owned by this process or another process.
+    QLocalSocket socket;
+    socket.connectToServer(mServerName);
+
+    if (socket.waitForConnected(WAIT_FOR_RESPONSE_MS)) {
+        const QString packagePathsData = mQueuedPackagePaths.join(QChar::LineFeed);
+        socket.write(packagePathsData.toUtf8());
+        socket.waitForBytesWritten(WAIT_FOR_RESPONSE_MS);
+        socket.disconnectFromServer();
+        return true;
+    }
+    return false;
+}
+
+
+// Attempt to start the Mudlet Server, return true if successful.
+// If a server was already running, cancel the attempt and return false.
+bool MudletInstanceCoordinator::tryToStart()
+{
+    QLocalSocket socket;
+    socket.connectToServer(mServerName);
+
+    // If there is no response within a reasonable time:
+    // Assume no server exists, and start a new server.
+    if (!socket.waitForConnected(WAIT_FOR_RESPONSE_MS)) {
+        listen(mServerName);
+        return true;
+    }
+    return false;
+}
+
+// Install all queued packages to a specified profile/host.
+// Mudlet will call this function whenever a profile is activated.
+void MudletInstanceCoordinator::installPackagesToHost(Host* activeProfile)
+{
+    mMutex.lock();
+    foreach(const QString& path, mQueuedPackagePaths) {
```

Comment:
Tiny detail but a modern C++ for range loop is preferred now, see https://github.com/KDE/clazy/blob/master/docs/checks/README-foreach.md.

As a side note, running https://github.com/KDE/clazy while developing Qt code is pretty useful.

Reply:
Thanks for catching that. I'll try to get clazy working.